druid/src/widget/view_switcher.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that can dynamically switch between one of many views.
+
+use crate::{
+    BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx,
+    Point, Rect, Size, UpdateCtx, Widget, WidgetPod,
+};
+
+/// A widget that can switch dynamically between one of many views depending
+/// application state.
+pub struct ViewSwitcher<T: Data, U: PartialEq> {
+    child_selector: Box<dyn Fn(&T, &Env) -> U>,
+    child_builder: Box<dyn Fn(&U, &Env) -> Box<dyn Widget<T>>>,
+    active_child: Option<WidgetPod<T, Box<dyn Widget<T>>>>,
+    active_child_id: Option<U>,
+}
+
+impl<T: Data, U: PartialEq> ViewSwitcher<T, U> {
+    /// Create a new view switcher.
+    ///
+    /// The `child_selector` closure is called every time the application data changes.
+    /// If the value it returns is the same as the one it returned during the previous
+    /// data change, nothing happens. If it returns a different value, then the
+    /// `child_builder` closure is called with the new value.
+    ///
+    /// The `child_builder` closure is expected to create a new child widget depending
+    /// on the selector to it.
+    pub fn new(
+        child_selector: impl Fn(&T, &Env) -> U + 'static,
+        child_builder: impl Fn(&U, &Env) -> Box<dyn Widget<T>> + 'static,
+    ) -> Self {
+        Self {
+            child_selector: Box::new(child_selector),
+            child_builder: Box::new(child_builder),
+            active_child: None,
+            active_child_id: None,
+        }
+    }
+}
+
+impl<T: Data, U: PartialEq> Widget<T> for ViewSwitcher<T, U> {
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut T, env: &Env) {
+        if let Some(ref mut child) = self.active_child {
```

Comment:
Another way of writing this is, `if let Some(child) = self.active_child.as_mut() {`. (either works, just wanted to point out the option)